## Optimizers

They are set of procedures defined in scipy that either find either the min value of a function or a root of the equation




## Optimizing Function:

All the algorithims which help in minimizing the data

In [ ]:
## Root of an  : x + cos(x).   we will solve it via optimzie.root function

## this function takes two arguments : "function_name" and x0       x0 is initial guess


from scipy.optimize import root
from math import cos

def eqn(x):
    return x+cos(x)


myroot=root(eqn,0)
print(myroot.x)


##eqn → function
## 0 → initial guess (x0)

[-0.73908513]


/var/folders/gg/f1jqwyw902s1lt078gd0wfdm0000gn/T/ipykernel_5338/2221896174.py:10: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return x+cos(x)


In [4]:
## here we want info not just about x but whole root 

print(myroot)

 message: The solution converged.
 success: True
  status: 1
     fun: [ 0.000e+00]
       x: [-7.391e-01]
  method: hybr
    nfev: 11
    fjac: [[-1.000e+00]]
       r: [-1.674e+00]
     qtf: [-2.668e-13]


message: The solution converged.

This is the solver’s explanation.

It means:

the algorithm successfully found a root iterations stopped correctly

success

success: True

Boolean value.

True → root found successfully
False → solver failed

Always check this first.

status

status: 1

Numerical exit code.

Common values:

1 → success

0 or negative → failure/warning

Think of it as a machine-readable version of message.

fun

fun: [0.000e+00]

Value of the function at the final solution.

You solved:

f(x)=x+cos(x)

At the root:

f(−0.73908513)≈0

So:

fun ≈ 0

Meaning:

the equation is satisfied

Scientific notation:

0.000e+00 = 0 × 10^0 = 0

x

x: [-7.391e-01]

This is the root (solution).

Scientific notation:

-7.391e-01
=
-7.391 × 10^{-1}
=
-0.7391

So the root is:

x≈−0.73908513

This satisfies:

x+cos(x)=0

method


method: hybr

Algorithm used internally.

hybr means:

Hybrid Powell Method

Combination of:

1-Newton-like methods

2-gradient approximations

Good for nonlinear equations.

This is SciPy’s default root solver.

nfev

nfev: 11

Number of function evaluations.

Meaning:

SciPy called your function eqn(x) 11 times.

Example internally:

eqn(0)
eqn(-0.5)
eqn(-0.7)
...

until convergence.

Smaller nfev:

usually faster
more efficient

fjac

fjac: [[-1.000e+00]]

Approximation of the Jacobian matrix.

For one variable:

J=dx/df
	​
For many variables:

Jij=∂xj / ∂fi

Since your function has only one variable:

f(x)=x+cos(x)

Derivative:

f′(x)=1−sin(x)

SciPy estimates this internally.

fjac stores factorized Jacobian information used by the solver.

For multivariable systems this becomes very important.

r

r: [-1.674e+00]

Part of QR factorization used internally.

The solver decomposes matrices like:

A=QR

where:

Q → orthogonal matrix
R → upper triangular matrix

r stores the R component.

Used for:

numerical stability
solving nonlinear systems efficiently

Mostly useful for advanced numerical analysis.

qtf

qtf: [-2.668e-13]

Represents:

QT * f

where:

QT→ transpose of orthogonal matrix
f → function vector

This measures how close the solver is to convergence.

Very tiny value:

−2.668×10^−13

means:

extremely close to exact root
convergence is excellent

## Minimizing the function

In [ ]:
## finding minima : we use scipy.optimize.minimize(). It uses three arguments
## arguments: "fun",x0,method
## method has some legal values : "CG","BFGS","NEWTON-CG","L-BFGS-B","TNC","COBYLA","SLSQP".

In [ ]:
## callback: functions called after each iteration of optimizations.

In [ ]:
from scipy.optimize import minimize

def equn(x):
    return x**2 + x + 2


mymin=minimize(equn,0,method="BFGS")
print(mymin.x)

[-0.50000001]


In [8]:
print(mymin)

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 1.75
        x: [-5.000e-01]
      nit: 2
      jac: [ 0.000e+00]
 hess_inv: [[ 5.000e-01]]
     nfev: 8
     njev: 4
